src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenUsingDeclarationTests.cs

```diff
@@ -1089,5 +1099,168 @@ static async Task Main()
             CompileAndVerify(source, expectedOutput: "Dispose; ");
         }
 
+        [Fact]
+        public void UsingDeclarationAsync()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+class C1 : IAsyncDisposable
+{
+    public ValueTask DisposeAsync() 
+    { 
+        Console.WriteLine(""Dispose async"");
+        return new ValueTask(Task.CompletedTask);
+    }
+}
+
+class C2
+{
+    static async Task Main()
+    {
+        await using C1 c = new C1();
+    }
+}";
+            var compilation = CreateCompilationWithTasksExtensions(source + _asyncDisposable, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(compilation, expectedOutput: "Dispose async").VerifyIL("C2.Main", @"
+{
+  // Code size       52 (0x34)
+  .maxstack  2
+  .locals init (C2.<Main>d__0 V_0,
+                System.Runtime.CompilerServices.AsyncTaskMethodBuilder V_1)
+  IL_0000:  newobj     ""C2.<Main>d__0..ctor()""
+  IL_0005:  stloc.0
+  IL_0006:  ldloc.0
+  IL_0007:  call       ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Create()""
+  IL_000c:  stfld      ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder C2.<Main>d__0.<>t__builder""
+  IL_0011:  ldloc.0
+  IL_0012:  ldc.i4.m1
+  IL_0013:  stfld      ""int C2.<Main>d__0.<>1__state""
+  IL_0018:  ldloc.0
+  IL_0019:  ldfld      ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder C2.<Main>d__0.<>t__builder""
+  IL_001e:  stloc.1
+  IL_001f:  ldloca.s   V_1
+  IL_0021:  ldloca.s   V_0
+  IL_0023:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start<C2.<Main>d__0>(ref C2.<Main>d__0)""
+  IL_0028:  ldloc.0
+  IL_0029:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder C2.<Main>d__0.<>t__builder""
+  IL_002e:  call       ""System.Threading.Tasks.Task System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Task.get""
+  IL_0033:  ret
+}");
+        }
+
+        [Fact]
+        public void UsingDeclarationAsyncPattern()
+        {
+            var source = @"
+using System.Threading.Tasks;
+class C1
+{
+    public Task DisposeAsync() 
+    { 
+        System.Console.WriteLine(""Dispose async"");
+        return Task.CompletedTask;
+    }
+}
+
+class C2
+{
+    static async Task Main()
+    {
+        await using C1 c = new C1();
```

Comment:
>new C1(); [](start = 27, length = 9)

Please add a Write in the constructor for C1, and add a Write below the `using`. This way the order of evaluation is made completely visible.